File: src/main/java/alfheim/common/spell/fire/SpellDispel.java
package alfheim.common.spell.fire;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import alexsocol.asjlib.ASJUtilities;
import alfheim.AlfheimCore;
import alfheim.api.entity.EnumRace;
import alfheim.api.spell.SpellBase;
import alfheim.client.render.world.SpellEffectHandlerClient.Spells;
import alfheim.common.core.handler.CardinalSystem.TargetingSystem;
import alfheim.common.core.handler.CardinalSystem.TargetingSystem.Target;
import alfheim.common.core.handler.SpellEffectHandler;
import alfheim.common.network.MessageEffect;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.potion.Potion;
import net.minecraft.potion.PotionEffect;

public class SpellDispel extends SpellBase {

	public SpellDispel() {
		super("dispel", EnumRace.SALAMANDER, 1000, 600, 25);
	}

	@Override
	public SpellCastResult performCast(EntityLivingBase caster) {
		if (!(caster instanceof EntityPlayer)) return SpellCastResult.NOTARGET; // TODO add targets for mobs
		
		Target tg = TargetingSystem.getTarget((EntityPlayer) caster);
		if (tg == null || tg.target == null) 
			return SpellCastResult.NOTARGET;
		
		if (tg.target != caster && !ASJUtilities.isInFieldOfVision(tg.target, caster)) return SpellCastResult.NOTSEEING;
			
		SpellCastResult result = checkCast(caster);
		if (result == SpellCastResult.OK) {
			List<PotionEffect> l = new ArrayList<PotionEffect>();
			for (Object o : tg.target.getActivePotionEffects())  if (Potion.potionTypes[((PotionEffect) o).getPotionID()].isBadEffect == tg.isParty) l.add((PotionEffect) o);
						
			if (l.isEmpty()) {
				tg.target.addPotionEffect(new PotionEffect(Potion.confusion.id, 300, 0, true));
				AlfheimCore.network.sendToAll(new MessageEffect(tg.target.getEntityId(), Potion.confusion.id, 300, 0));
			} else {
				for (PotionEffect pe : l) {
					tg.target.removePotionEffect(pe.potionID);
					AlfheimCore.network.sendToAll(new MessageEffect(tg.target.getEntityId(), pe.getPotionID(), 0, 0));
				}
			}
			SpellEffectHandler.sendPacket(Spells.DISPEL, tg.target);
		}
		
		return result;
	}
}

File: src/main/java/alfheim/common/spell/fire/SpellDispel.kt
package alfheim.common.spell.fire

import alexsocol.asjlib.ASJUtilities
import alfheim.AlfheimCore
import alfheim.api.entity.EnumRace
import alfheim.api.spell.SpellBase
import alfheim.client.render.world.VisualEffectHandlerClient.VisualEffects
import alfheim.common.core.handler.*
import alfheim.common.core.handler.CardinalSystem.TargetingSystem
import alfheim.common.security.InteractionSecurity
import alfheim.common.network.MessageEffect
import net.minecraft.entity.EntityLivingBase
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.potion.*
import java.util.*

object SpellDispel: SpellBase("dispel", EnumRace.SALAMANDER, 1000, 600, 25) {
	
	override var duration = 300
	
	override val usableParams: Array<Any>
		get() = arrayOf(duration)
	
	override fun performCast(caster: EntityLivingBase): SpellCastResult {
		if (caster !is EntityPlayer) return SpellCastResult.NOTARGET // TODO add targets for mobs
		
		val tg = TargetingSystem.getTarget(caster)
		val tgt = tg.target ?: return SpellCastResult.NOTARGET
		
		if (tgt !== caster && ASJUtilities.isNotInFieldOfVision(tgt, caster)) return SpellCastResult.NOTSEEING
		
		if (!InteractionSecurity.canDoSomethingWithEntity(caster, tgt)) return SpellCastResult.NOTALLOW
		
		val result = checkCast(caster)
		if (result == SpellCastResult.OK) {
			val l = ArrayList<PotionEffect>()
			for (o in tgt.activePotionEffects) if (Potion.potionTypes[(o as PotionEffect).potionID].isBadEffect == tg.isParty) if (o.potionID != AlfheimConfigHandler.potionIDLeftFlame) l.add(o)
			
			if (l.isEmpty()) {
				tgt.addPotionEffect(PotionEffect(Potion.confusion.id, duration, 0, true))
				AlfheimCore.network.sendToAll(MessageEffect(tgt.entityId, Potion.confusion.id, duration, 0))
			} else {
				for (pe in l) {
					tgt.removePotionEffect(pe.potionID)
					AlfheimCore.network.sendToAll(MessageEffect(tgt.entityId, pe.getPotionID(), 0, 0))
				}
			}
			VisualEffectHandler.sendPacket(VisualEffects.DISPEL, tgt)
		}
		
		return result
	}
}